### PR TITLE
Fix #10129: Fix comparisons of applied type aliases

### LINKED ISSUE #10129: Aliases to Hard Unions don't reduce
## Minimized code

When we make an alias to a union type, it is not reduced compared to a raw union type, i.e. in the following `Lifted[Int]` is not the same as `Err | Int`, despite being an alias

```Scala
class Err                                                                                                        

type Lifted[A] = Err | A

def point[O](o: O): Lifted[O] = o
extension [O, U](o: Lifted[O]) def map(f: O => U): Lifted[U] = ???

val error: Err = Err()

def ok: Int | Err =
  point("a").map(_ => if true then 1 else error)

def fail: Lifted[Int] =
  point("a").map(_ => if true then 1 else error) // error
```

## Output

```scala
14 |  point("a").map(_ => if true then 1 else error) // error
   |                                          ^^^^^
   |                                          Found:    (error : Err)
   |                                          Required: Int
```

## Expectation

both `ok` and `fail` should compile

### PR DESCRIPTION
**Pull request by odersky**
Given

    type F[Xs] = ...
    F[As] <: F[Bs]

Do we dealias `F` or check the arguments directly? It turns out that neither choice
is correct in all instances. Checking the arguments directly is necessary for hk
type inference, but may narrow the constraint too much. The solution is to use
an `either` that tries both alternatives.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -997,6 +997,25 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
...
+          // An example is i10129.scala. Here, we have the following situation:
+          //
+          //    type Lifted[A] = Err | A
+          //    def point[O](o: O): Lifted[O] = o
+          //    extension [O, U](o: Lifted[O]) def map(f: O => U): Lifted[U] = ???
+          //    point("a").map(_ => if true then 1 else error)
+          //
+          // This leads to the constraint Lifted[U] <: Lifted[Int]. If we just
+          // check the arguments this gives `U <: Int`. But this is wrong. Dealiasing
+          // `Lifted` gives `Err | U <: Err | Int`, hence it should be `U <: Err | Int`.
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1010]: I'm not necessarily against changing it now if we believe it's worth it, but that was actually an intentional behavior, see the justification in https://github.com/lampepfl/dotty/commit/5fb13aad58cc554e7afc3b15f69d4d16698a1e5f:

> I think the new behavior is arguably better since using type aliases now
gives you more control on how type inference proceeds, even if it means
that some things that used to typecheck don't anymore.

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1010]: I think we should not simply infer something that does not follow from the constraint. Or at least we should do it under the strict control of `either`. i10129 is arguably a very natural way to express things.

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1010]: Or, let's put it that way. The _correct_ thing to do is to dealias. Unfortunately that breaks hk type inference. But hk type inference always was an ad-hoc best effort thing (some might call it a kludge). It's clear we can't do hk type inference in general. So it's good to push this as far as we can, but it should not compromise the correct behavior for other constraints.

**Comment by odersky** [comment]
test performance please

**Review by smarter — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 2]
```diff
@@ -1039,9 +1058,35 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
...
+                                return recur(tp1a, tp2a)
+                              case _ =>
+                          case _ =>
+                        // Otherwise either compare arguments or compare dealiased types.
+                        // Note: This test should be done with an `either`, but this fails
+                        // for hk-alias-unification.scala The problem is not GADT boundschecking;
+                        // that is taken care of by the outer inFrozenGadtIf test. The problem is
+                        // that hk-alias-unification.scala relies on the right isSubArgs being performed
+                        // when constraining the result type of a method. But there we are only allowed
+                        // to use a necessaryEither, since it might be that an implicit conversion will
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1077]: You are right. The change was actually introduced here: https://github.com/lampepfl/dotty/pull/8635

It was a way to avoid overcommitting when constraining the result type. Unfortunately, it seems that this causes the problems we encounter here. So, what to do? All this is getting more and more ad-hoc, which is not very satisfactory.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 3]
```diff
@@ -1039,9 +1058,35 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
...
+                        // be inserted on the result. So strictly speaking by going to a sufficientEither
+                        // we overshoot and narrow the constraint unnecessarily. On the other hand, this
+                        // is probably an extreme corner case. If we exclude implicit conversions, the
+                        // problem goes away since in that case we would have a normal subtype test here,
+                        // which demands a sufficientEither anyway.
+                        sufficientEither(
+                          isSubArgs(args1, args2, tp1, tparams),
+                          recur(tp1.superType, tp2.superType))
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1085]: The code is only hit if the two type constructors are the same.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 4]
```diff
@@ -1039,9 +1058,35 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
...
-                      isSubArgs(args1, args2, tp1, tparams)
-                    }
+                    def checkArgs(): Boolean =
+                      if tycon1sym == tycon2sym && tycon1sym.isAliasType then
+                        // if types are simple forwarding aliases then try the dealiased types
+                        tp1.superType match
+                          case tp1a as AppliedType(_, args1a) if args1a.eqElements(args1) =>
+                            tp2.superType match
+                              case tp2a as AppliedType(_, args2a) if args2a.eqElements(args2) =>
+                                return recur(tp1a, tp2a)
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1068]: Ah, good to know!

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 5]
```diff
@@ -1039,9 +1058,35 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
...
+                          case _ =>
+                        // Otherwise either compare arguments or compare dealiased types.
+                        // Note: This test should be done with an `either`, but this fails
+                        // for hk-alias-unification.scala The problem is not GADT boundschecking;
+                        // that is taken care of by the outer inFrozenGadtIf test. The problem is
+                        // that hk-alias-unification.scala relies on the right isSubArgs being performed
+                        // when constraining the result type of a method. But there we are only allowed
+                        // to use a necessaryEither, since it might be that an implicit conversion will
+                        // be inserted on the result. So strictly speaking by going to a sufficientEither
+                        // we overshoot and narrow the constraint unnecessarily. On the other hand, this
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1079]: I believe not, since we are in a frozen Gadt state anyway.

[the remark this answers is not in the document]
Comment (smarter) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1079]: Ah, we're in `inFrozenGadtIf(!tyconIsInjective)`, and `tyconIsInjective` won't be true if both constructors are type aliases, so gadt constraints are frozen indeed, although it's not immediately obvious from the code.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 6]
```diff
@@ -1031,11 +1056,34 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
...
+                        // So strictly speaking by going to a sufficientEither we overshoot and narrow
+                        // the constraint unnecessarily. But unlike in the situation of #8365 with test cases
+                        // and-inf.scala and or-inf.scala, we don't cut off solutions in principle by dealiasing.
+                        // We "just" risk missing an opportunity to do a desired hk-type inference. This
+                        // is hopefully a less frequent corner case. Where it goes wrong compared to the status
+                        // before this fix: If isSubArgs succeeds and then the dealised test also succeeds
+                        // while constraining strictly less than the isSubArgs, then the we used to pick the
+                        // constraint after isSubArgs, but we now pick the constraint after the dealiased test.
+                        // There's one test in shapeless/deriving that had to be disabled
+                        // (modules/deriving/src/test/scala/shapeless3/deriving/deriving.scala, value v7 in the functor test).
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1079]: FYI @milessabin, how important is that particular test?

**Comment by odersky** [comment]
That's the second time I get a conflict because of shapeless which causes me a huge effort to re-sync. I'll do that one last time and then merge since I don't have the time to go through these steps a third time.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala (hunk 6 above)
Comment (milessabin) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:1079]: That's an important test. It demonstrates the ability to derive functor instances for nested type constructors.

It would at least be good to know why `v7` fails whereas `v6` succeeds. Both involve recursion. The only difference is that the former failing one has an additional non-recursive wrapper, which is itself successfully derived in `v3`.

**Comment by milessabin** [comment]
The failure with shapeless here is demonstrating what this change is forcing us to give up.

That being the case, it would help to have a motivation for the change and a sketch of a workaround.

**Comment by odersky** [comment]
The motivation is issue #10129. 

Here we have
```scala
type Lifted[A] = Err | A
```
and the constraint
```scala
Lifted[X] <: Lifted[Int]
```
It's simply incorrect to conclude from this that `X <: Int`. So that means in general we have to dealias before comparing. But doing this breaks some higher-kinded type inference moves. So the compromise in this PR is to look at both the direct arguments and the dealiased type and pick the solution that constrains less. That still allows hk type inference to work, unless the dealiased
comparison would constrain things strictly less. 

I don't know what else one could do here.

**Comment by milessabin** [comment]
That sounds completely reasonable.

What's not clear to me is why that breaks the test in shapeless. Other than involving hk types it doesn't appear to be related in any obvious why the the example you're fixing here.

You suggest this is an inference problem ... is there an explicit type annotation or parameter which could be added to the test which would fix it? What is the compilation error in shapeless for that test?

**Comment by odersky** [comment]
> What's not clear to me is why that breaks the test in shapeless. Other than involving hk types it doesn't appear to be related in any obvious way to the example you're fixing here.

It's a hard to diagnose since it is in auto-generated code form deriving, I believe. What must have happened is that there is a type alias somewhere such as 
```scala
    type T[X] = Any
```
and a constraint gets generated such as
```scala
    T[F[A]]  <: T[List{X]]
```
In this case we'd conclude so far that `F = List`. But after dealiasing it's just `T[F[A]] <: Any` which does not constrain at all. So we do not instantiate `F`. 

The RHS of `Any` is just for illustration, it could be anything that is already a supertype of the lhs `T[F[A]]`.

**Comment by milessabin** [comment]
>  it is in auto-generated code from deriving

It's in completely vanilla (albeit sophisticated) code. Match types and inlining.

**Comment by milessabin** [comment]
There are uses of a `Const` type constructor in shapeles, ie.
```scala
type Const[c] = [t] =>> c
```
which could produce something similar to your `type T[X] = Any` (with something other than `Any` on the RHS).

Are you saying that this sort of construct can't be relied on any more? Can you suggest an alternative encoding?

**Comment by odersky** [comment]
Yes, `Const` would be problematic for the same reasons as my example.Not sure what to do. I imagine `Const` can't be opaque. right?

**Comment by smarter** [comment]
I think this comes down to a design decision: do we want type aliases to be a way to drive type inference, or do we want to always infer the minimum set of constraints? The former let library authors craft aliases to improve inference in their APIs and hopefully provide a better UX, but it's ill-defined, whereas the latter can be unintuitive for users but it's a clear criterion compiler developers can refer to when making changes to type inference, so it reduces the risk that we end up piling up heuristics after heuristics without really knowing what we're doing.
I don't know what the correct answer is here, but given that using type aliases to drive inference is already well-established with HK unification which works fairly well in practice, I don't really mind that `Lifted[?U] <: Lifted[A]` infers `?U <: A` no matter what `Lifted` is.

**Comment by milessabin** [comment]
> I imagine `Const` can't be opaque. right?

No, it can't.

**Comment by smarter** [comment]
A trick might be to write `type Const[c] >: [t] =>> c <: [t] =>> c` so the compiler isn't so eager to get rid of the alias (I think we wouldn't end up triggering the code added in this PR because `isAliasType` will return false), but even if it works it's not a very nice solution.

**Comment by odersky** [comment]
```scala
Const[c] >: [t] =>> c <: [t] =>> c
```
Yes, that looks promising.

**Comment by odersky** [comment]
How should we proceed? We don't seem to have an alternative to the fix in this PR. I propose we get it in and then see what needs to be done to bring back the failing test in shapeless.

**Comment by milessabin** [comment]
I'm still unclear why this test fails but others don't.

`Const` is used pervasively in shapeless, and there's a presumption in the Dotty type class derivation infrastructure that something like `Const` be workable. If this is an edge case with a workaround it's not a big deal, but then it would help to understand _why_ this is an edge case whereas other uses of `Const` aren't. If it isn't then there's a chance that this will damage type class derivation across the board.

**Comment by odersky** [comment]
Good news! I found a way to do what we want and still keep all shapeless tests. 

Digging deeper I found that the issue was not what I initially suspected. I suspected that a constraint generated by a isSubArgs was rolled back by the test of the dealiased types. But what happened instead was that we had a failing `isSubArgs`, and the alternative that recurred on the dealiased types somehow missed something which then led to this error:
```
-- Error: deriving.scala:104:43 ------------------------------------------------
104 |    val v7 = Functor[[t] =>> CList[Opt[t]]]
    |                                           ^
    |ambiguous implicit arguments: both method functorGen in object Functor and method given_Functor_F in object Functor match type shapeless3.deriving.Functor[
    |  [t] =>> shapeless3.deriving.adts.CList[shapeless3.deriving.adts.Opt[t]]
    |] of parameter ff of method apply in object Functor
1 error found
```
If we simply fail `isMatchingApply` on a failed `isSubArgs` then the test compiles. So this means that the different paths we took for checking the dealiased types somehow made a difference. In the failing example we dealised both types in lockstep whereas in the original code we dealiased first the supertype and then went through a second check via `compareApplied1`. Presumably that second check could have done some hk inference that we missed by dealiasing both types at the same time. 

The fix is simple: Fail is `isSubArgs` fails. If it succeeds, also check the dealiased types and pick the weaker constraint.

**Comment by milessabin** [comment]
> Good news! I found a way to do what we want and still keep all shapeless tests.

Awesome ... thanks for digging in to this! :smile:
